lib/nat/service.go

```diff
@@ -206,41 +206,39 @@ func (s *Service) RemoveMapping(mapping *Mapping) {
 // Optionally takes renew flag which indicates whether or not we should renew
 // mappings with existing natds
 func (s *Service) updateMapping(ctx context.Context, mapping *Mapping, nats map[string]Device, renew bool) {
-	var added, removed []Address
-
 	renewalTime := time.Duration(s.cfg.Options().NATRenewalM) * time.Minute
 	mapping.expires = time.Now().Add(renewalTime)
 
-	newAdded, newRemoved := s.verifyExistingMappings(ctx, mapping, nats, renew)
-	added = append(added, newAdded...)
-	removed = append(removed, newRemoved...)
+	change := false
+	func() {
+		mapping.mut.Lock()
+		defer mapping.mut.Unlock()
 
-	newAdded, newRemoved = s.acquireNewMappings(ctx, mapping, nats)
-	added = append(added, newAdded...)
-	removed = append(removed, newRemoved...)
+		change = s.verifyExistingMappings(ctx, mapping, nats, renew)
+		add := s.acquireNewMappings(ctx, mapping, nats)
+		change = change || add
+	}()
```

Comment:
The intention here is to be able to `defer` the lock - or is there more to this function? Otherwise that seems a bit unnecessary to mean, compared to lock-unlock unclosing just 2-3 "plan" lines (no control flow shenanigans).

Reply:
Yes, I wanted to defer the unlock so it happens even when something panics (can happen for out-of-memory, in theory). Reverted to a simple lock-change-unlock.